**About Book Crossing Dataset**<br>

This dataset has been compiled by Cai-Nicolas Ziegler in 2004, and it comprises of three tables for users, books and ratings. Explicit ratings are expressed on a scale from 1-10 (higher values denoting higher appreciation) and implicit rating is expressed by 0.

Reference: http://www2.informatik.uni-freiburg.de/~cziegler/BX/ 

**Objective**

This project entails building a Book Recommender System for users based on user-based and item-based collaborative filtering approaches.

#### Execute the below cell to load the datasets

In [1057]:
import pandas as pd
import numpy as np

In [1058]:
#Loading data
books = pd.read_csv("books.csv", sep=";", error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']

users = pd.read_csv('users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']

ratings = pd.read_csv('ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.p

### Check no.of records and features given in each dataset

In [1059]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
ISBN                 271360 non-null object
bookTitle            271360 non-null object
bookAuthor           271359 non-null object
yearOfPublication    271360 non-null object
publisher            271358 non-null object
imageUrlS            271360 non-null object
imageUrlM            271360 non-null object
imageUrlL            271357 non-null object
dtypes: object(8)
memory usage: 16.6+ MB


## book dataset contains about 271360 records with 8 features.

In [1060]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
userID      278858 non-null int64
Location    278858 non-null object
Age         168096 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


users dataset contains about 278858 records with 3 features.

In [1061]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
userID        1149780 non-null int64
ISBN          1149780 non-null object
bookRating    1149780 non-null int64
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


ratings dataset contains about 1149780 records with 3 features.

## Exploring books dataset

In [1062]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

### Drop last three columns containing image URLs which will not be required for analysis

In [1063]:
books.drop(books.iloc[:,-3:],inplace=True,axis=1)

In [1064]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  
0    Mark P. O. Morford              2002     Oxford University Press  
1  Richard Bruce Wright              2001       HarperFlamingo Canada  
2          Carlo D'Este              1991             HarperPerennial  
3      Gina Bari Kolata              1999        Farrar Straus Giroux  
4       E. J. W. Barber              1999  W. W. Norton &amp; Company

**yearOfPublication**

### Check unique values of yearOfPublication


In [1065]:
yeatOfPublication = books['yearOfPublication'].unique()

In [1066]:
len(yeatOfPublication)

202

In [1067]:
yeatOfPublication

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

# As it can be seen from above that there are some incorrect entries in this field. It looks like Publisher names 'DK Publishing Inc' and 'Gallimard' have been incorrectly loaded as yearOfPublication in dataset due to some errors in csv file.


Also some of the entries are strings and same years have been entered as numbers in some places. We will try to fix these things in the coming questions.

### Check the rows having 'DK Publishing Inc' as yearOfPublication

In [1068]:
books.loc[books['yearOfPublication'] == 'DK Publishing Inc']

ISBN                                          bookTitle  \
209538  078946697X  DK Readers: Creating the X-Men, How It All Beg...   
221678  0789466953  DK Readers: Creating the X-Men, How Comic Book...   

       bookAuthor  yearOfPublication  \
209538       2000  DK Publishing Inc   
221678       2000  DK Publishing Inc   

                                                publisher  
209538  http://images.amazon.com/images/P/078946697X.0...  
221678  http://images.amazon.com/images/P/0789466953.0...

### Drop the rows having `'DK Publishing Inc'` and `'Gallimard'` as `yearOfPublication`

In [1069]:
books_df=books[books.yearOfPublication != 'DK Publishing Inc']
books_df=books_df[books_df.yearOfPublication != 'Gallimard']

In [1070]:
books_df.loc[books_df['yearOfPublication'] == 'DK Publishing Inc']

Empty DataFrame
Columns: [ISBN, bookTitle, bookAuthor, yearOfPublication, publisher]
Index: []

In [1071]:
books_df.loc[books_df['yearOfPublication'] == 'Gallimard']

Empty DataFrame
Columns: [ISBN, bookTitle, bookAuthor, yearOfPublication, publisher]
Index: []

### Change the datatype of yearOfPublication to 'int'

In [1072]:
books_df['yearOfPublication'] = pd.to_numeric(books_df['yearOfPublication'])

In [1073]:
books_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 271357 entries, 0 to 271359
Data columns (total 5 columns):
ISBN                 271357 non-null object
bookTitle            271357 non-null object
bookAuthor           271356 non-null object
yearOfPublication    271357 non-null int64
publisher            271355 non-null object
dtypes: int64(1), object(4)
memory usage: 12.4+ MB


### Drop NaNs in `'publisher'` column


In [1074]:
books_df['publisher'].dropna(inplace=True)

In [1075]:
books_df['publisher'].isna().sum()

0

## Exploring Users dataset

In [1076]:
print(users.shape)
users.head()

(278858, 3)


userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

### Get all unique values in ascending order for column `Age`

In [1077]:
userAge = users['Age'].unique()

In [1078]:
len(userAge)

166

In [1079]:
user_sorted = users.sort_values(by=['Age'])
user_sorted['Age'].unique()

array([  0.,   1.,   2.,   3.,   4.,   5.,   6.,   7.,   8.,   9.,  10.,
        11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,  19.,  20.,  21.,
        22.,  23.,  24.,  25.,  26.,  27.,  28.,  29.,  30.,  31.,  32.,
        33.,  34.,  35.,  36.,  37.,  38.,  39.,  40.,  41.,  42.,  43.,
        44.,  45.,  46.,  47.,  48.,  49.,  50.,  51.,  52.,  53.,  54.,
        55.,  56.,  57.,  58.,  59.,  60.,  61.,  62.,  63.,  64.,  65.,
        66.,  67.,  68.,  69.,  70.,  71.,  72.,  73.,  74.,  75.,  76.,
        77.,  78.,  79.,  80.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,
        88.,  89.,  90.,  91.,  92.,  93.,  94.,  95.,  96.,  97.,  98.,
        99., 100., 101., 102., 103., 104., 105., 106., 107., 108., 109.,
       110., 111., 113., 114., 115., 116., 118., 119., 123., 124., 127.,
       128., 132., 133., 136., 137., 138., 140., 141., 143., 146., 147.,
       148., 151., 152., 156., 157., 159., 162., 168., 172., 175., 183.,
       186., 189., 199., 200., 201., 204., 207., 20

Age column has some invalid entries like nan, 0 and very high values like 100 and above

### Values below 5 and above 90 do not make much sense for our book rating case...hence replace these by NaNs

In [1080]:
user_sorted.loc[user_sorted['Age']<5,'Age'] = np.NaN

In [1081]:
user_sorted.loc[user_sorted['Age']>90,'Age'] = np.NaN

In [1082]:
user_sorted.Age.unique()

array([nan,  5.,  6.,  7.,  8.,  9., 10., 11., 12., 13., 14., 15., 16.,
       17., 18., 19., 20., 21., 22., 23., 24., 25., 26., 27., 28., 29.,
       30., 31., 32., 33., 34., 35., 36., 37., 38., 39., 40., 41., 42.,
       43., 44., 45., 46., 47., 48., 49., 50., 51., 52., 53., 54., 55.,
       56., 57., 58., 59., 60., 61., 62., 63., 64., 65., 66., 67., 68.,
       69., 70., 71., 72., 73., 74., 75., 76., 77., 78., 79., 80., 81.,
       82., 83., 84., 85., 86., 87., 88., 89., 90.])

### Replace null values in column `Age` with mean

In [1083]:
age_mean = user_sorted['Age'].mean()

In [1084]:
user_sorted['Age'] = user_sorted['Age'].replace(np.NaN,age_mean)

In [1085]:
user_sorted.head()

userID                                Location       Age
162239  162240                  buffalo, new york, usa  34.72384
130115  130116  lower templestowe, victoria, australia  34.72384
135953  135954            powder springs, georgia, usa  34.72384
207490  207491                 manassas, virginia, usa  34.72384
189924  189925          san francisco, california, usa  34.72384

### Change the datatype of `Age` to `int`

In [1086]:
user_sorted['Age'] = user_sorted['Age'].astype(int)

In [1087]:
print(sorted(user_sorted.Age.unique()))

[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90]


## Exploring the Ratings Dataset

### check the shape

In [1088]:
ratings.shape

(1149780, 3)

In [1089]:
n_users = users.shape[0]
n_books = books.shape[0]
print("No of Users:",n_users)
print("No of Books:",n_books)

No of Users: 278858
No of Books: 271360


### Ratings dataset should have books only which exist in our books dataset. Drop the remaining rows

In [1090]:
ratings = ratings[ratings['ISBN'].isin(books['ISBN'])]

In [1091]:
ratings.shape

(1031136, 3)

We have taken only the rating records for which the books are available in books dataset

### Ratings dataset should have ratings from users which exist in users dataset. Drop the remaining rows

In [1092]:
ratings = ratings[ratings['userID'].isin(users['userID'])]

In [1093]:
ratings.shape

(1031136, 3)

Looks likes there are no records removed, which signifies all user in ratings dataset existing in users dataset as well.

### Consider only ratings from 1-10 and leave 0s in column `bookRating`

In [1094]:
ratings = ratings[ratings['bookRating'].between(1,10)]

In [1095]:
ratings.shape

(383842, 3)

### Find out which rating has been given highest number of times

In [1096]:
ratings['bookRating'].mode()

0    8
dtype: int64

Rating 8 is given more number of times.

### **Collaborative Filtering Based Recommendation Systems**

### For more accurate results only consider users who have rated atleast 100 books

In [1097]:
ratings_filtered = ratings.groupby('userID').filter(lambda x: len(x) >= 100)

In [1098]:
ratings_filtered.shape

(103271, 3)

### Generating ratings matrix from explicit ratings


In [1099]:
from sklearn.model_selection import train_test_split

In [1100]:
trainDF, tempDF = train_test_split(ratings_filtered, test_size=0.2, random_state=100)
#creating a copy of tempDF as testDF
testDF = tempDF.copy()
#Assigning ratings of tempDF to nan
tempDF.bookRating = np.nan

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [1101]:
#Creating ratings with trainDF and tempDF
noRatings_df= pd.concat([trainDF, tempDF]).reset_index()

In [1102]:
noRatings_df.head()

index  userID        ISBN  bookRating
0  426442  101851  0890878218        10.0
1   48772   11676  0385474016        10.0
2  284111   67840  019285304X        10.0
3  410090   98391  0756400503         9.0
4  409008   98391  0451459024         9.0

The format of matrix is going to be like each user in a row and one column per ISBN ID.

In [1103]:
rf_df = noRatings_df.pivot(index = 'userID', columns ='ISBN', values = 'bookRating').fillna(0)
rf_df.tail()

ISBN    0000913154  0001046438  000104687X  0001047213  0001047973  \
userID                                                               
274061         0.0         0.0         0.0         0.0         0.0   
274301         0.0         0.0         0.0         0.0         0.0   
275970         0.0         0.0         0.0         0.0         0.0   
277427         0.0         0.0         0.0         0.0         0.0   
278418         0.0         0.0         0.0         0.0         0.0   

ISBN    000104799X  0001048082  0001053736  0001053744  0001055607  \
userID                                                               
274061         0.0         0.0         0.0         0.0         0.0   
274301         0.0         0.0         0.0         0.0         0.0   
275970         0.0         0.0         0.0         0.0         0.0   
277427         0.0         0.0         0.0         0.0         0.0   
278418         0.0         0.0         0.0         0.0         0.0   

ISBN       ...      B000092Q0A  B00009EF82  B00009NDAN  B0000DYXID  \
userID     ...                                                       
274061     ...             0.0         0.0         0.0         0.0   
274301     ...             0.0         0.0        10.0         0.0   
275970     ...             0.0         0.0         0.0         0.0   
277427     ...             0.0         0.0         0.0         0.0   
278418     ...             0.0         0.0         0.0         0.0   

ISBN    B0000T6KHI  B0000VZEJQ  B0000X8HIE  B00013AX9E  B0001I1KOG  B000234N3A  
userID                                                                          
274061         0.0         0.0         0.0         0.0         0.0         0.0  
274301         0.0         0.0         0.0         0.0         0.0         0.0  
275970         0.0         0.0         0.0         0.0         0.0         0.0  
277427         0.0         0.0         0.0         0.0         0.0         0.0  
278418         0.0         0.0         0.0         0.0         0.0         0.0  

[5 rows x 66574 columns]

### Generate the predicted ratings using SVD with no.of singular values to be 50

In [1104]:
from scipy.sparse.linalg import svds

In [1105]:
U, sigma, Vt = svds(rf_df, k = 50)

In [1106]:
#diag
sigma = np.diag(sigma)

In [1107]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) 
preds_df = pd.DataFrame(all_user_predicted_ratings, columns = rf_df.columns)

### Take a particular user_id

### Lets find the recommendations for user with id `2110`

#### Note: Execute the below cells to get the variables loaded

In [1108]:
userID = 2110

In [1109]:
user_id = 2 #2nd row in ratings matrix and predicted matrix

### Get the predicted ratings for userID `2110` and sort them in descending order

In [1110]:
user_row_number = user_id - 1 # UserID starts at 1, not 0
sorted_user_predictions = preds_df.iloc[user_row_number].sort_values(ascending=False)

In [1111]:
sorted_user_predictions.head()

ISBN
0451156609    0.182518
0345368932    0.177451
0451180232    0.176427
0451167317    0.173647
0671733893    0.167263
Name: 1, dtype: float64

### Create a dataframe with name `user_data` containing userID `2110` explicitly interacted books

In [1112]:
user_data = ratings[ratings.userID==userID]

In [1113]:
user_data.head()

userID        ISBN  bookRating
14448    2110  0060987529           7
14449    2110  0064472779           8
14450    2110  0140022651          10
14452    2110  0142302163           8
14453    2110  0151008116           5

In [1114]:
user_data.shape

(103, 3)

### Combine the user_data and and corresponding book data(`book_data`) in a single dataframe with name `user_full_info`

In [1115]:
book_data = books[books.ISBN.isin(user_data.ISBN)]

In [1116]:
book_data.shape

(103, 5)

In [1117]:
book_data.head()

ISBN                                          bookTitle  \
246   0151008116                                         Life of Pi   
904   015216250X  So You Want to Be a Wizard: The First Book in ...   
1000  0064472779                                  All-American Girl   
1302  0345307674                     Return of the Jedi (Star Wars)   
1472  0671527215                  Hitchhikers's Guide to the Galaxy   

         bookAuthor yearOfPublication           publisher  
246     Yann Martel              2002            Harcourt  
904     Diane Duane              2001  Magic Carpet Books  
1000      Meg Cabot              2003        HarperTrophy  
1302     James Kahn              1983       Del Rey Books  
1472  Douglas Adams              1984              Pocket

In [1118]:
 user_full_info = (user_data.merge(book_data, how = 'left', left_on = 'ISBN', right_on = 'ISBN').
                     sort_values(['bookRating'], ascending=False)
                 )

In [1119]:
user_full_info.head()

userID        ISBN  bookRating  \
76    2110  067166865X          10   
52    2110  0590109715          10   
64    2110  0590629786          10   
63    2110  0590629778          10   
61    2110  059046678X          10   

                                            bookTitle           bookAuthor  \
76  STAR TREK YESTERDAY'S SON (Star Trek: The Orig...         A.C. Crispin   
52  The Andalite Chronicles (Elfangor's Journey, A...  Katherine Applegate   
64                      The Visitor (Animorphs, No 2)      K. A. Applegate   
63                     The Invasion (Animorphs, No 1)      K. A. Applegate   
61                                       The Yearbook       Peter Lerangis   

   yearOfPublication   publisher  
76              1988  Audioworks  
52              1997       Apple  
64              1996  Scholastic  
63              1996  Scholastic  
61              1994  Scholastic

### Get top 10 recommendations for above given userID from the books not already rated by that user

In [1120]:
print ('User {0} has already rated {1} movies.'.format(userID, user_full_info.shape[0]))

User 2110 has already rated 103 movies.


Recommending the highest 10 predicted ratings movies not already rated.

In [1121]:
recommendations = (books[~books['ISBN'].isin(ratings['ISBN'])].
merge(pd.DataFrame(sorted_user_predictions).reset_index(), how = 'left', left_on = 'ISBN', right_on = 'ISBN').
rename(columns = {user_row_number: 'Predictions'}))

In [1122]:
#sort the prediction values for 10 recommendations
recommendations = (recommendations.sort_values('Predictions', ascending = False).
                    iloc[:10, :-1])

In [1123]:
recommendations

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0393045218                             The Mummies of Urumchi   
2  0679425608  Under the Black Flag: The Romance and the Real...   
3  080652121X  Hitler's Secret Bankers: The Myth of Swiss Neu...   
4  3442353866    Der Fluch der Kaiserin. Ein Richter- Di- Roman.   
5  3442410665                                  Sturmzeit. Roman.   
6  3442446937                                 Tage der Unschuld.   
7  055321215X                                Pride and Prejudice   
8  0812523873                                            Proxies   
9  0553280333                                 Getting Well Again   

                bookAuthor yearOfPublication                   publisher  
0       Mark P. O. Morford              2002     Oxford University Press  
1          E. J. W. Barber              1999  W. W. Norton &amp; Company  
2          David Cordingly              1996                Random House  
3               Adam Lebor              2000               Citadel Press  
4           Eleanor Cooney              2001                    Goldmann  
5           Charlotte Link              1991                    Goldmann  
6  Richard North Patterson              2000                    Goldmann  
7              Jane Austen              1983                      Bantam  
8           Laura J. Mixon              1999                   Tor Books  
9     O. Carol Simonton Md              1992                      Bantam